edb/edgeql/compiler/stmt.py

```diff
@@ -128,8 +128,18 @@ def compile_ForQuery(
 
         ictx = iterator_ctx or sctx
 
+        contains_dml = qlutils.contains_dml(qlstmt.result)
+        # If the body contains DML, then we need to prohibit
+        # correlation between the iterator and the enclosing
+        # query, since the correlation imposes compilation issues
+        # we aren't willing to tackle.
+        if contains_dml:
+            ictx.path_scope.factoring_allowlist.update(
+                ctx.iterator_path_ids)
         iterator_view = stmtctx.declare_view(
             iterator, qlstmt.iterator_alias,
+            factoring_fence=contains_dml,
+            temporary=False,
```

Comment:
The underlying cause seems to be a bug in `attach_subtree`.  Removing `temporary` here helps in this particular case, but the bug is still there.  Here's an example that's broken and doesn't use `FOR`:

```
WITH MODULE test
SELECT count(
    (Card.name, (SELECT (SELECT Card)))
)
```

Reply:
Per slack discussion, the real broken example is
```
        WITH MODULE test
        SELECT (
            Card.name,
            (WITH X := (SELECT Card) SELECT X.name),
        )
```